File: aws/src/handlers/auth_handler.py
import json
import boto3
import os
import uuid
from datetime import datetime
from botocore.exceptions import ClientError
from cors_utils import cors_response, cors_error_response, handle_preflight_request, extract_origin_from_event

# Initialize AWS clients
cognito = boto3.client('cognito-idp')
dynamodb = boto3.resource('dynamodb')

# Environment variables
USER_POOL_ID = os.environ['COGNITO_USER_POOL_ID']
CLIENT_ID = os.environ['COGNITO_CLIENT_ID']
USERS_TABLE = os.environ['DYNAMODB_TABLE_USERS']

users_table = dynamodb.Table(USERS_TABLE)

def lambda_handler(event, context):
    """Main Lambda handler for authentication operations"""
    try:
        http_method = event['httpMethod']
        path = event['path']
        origin = extract_origin_from_event(event)
        
        print(f"Processing {http_method} {path} from origin: {origin}")
        
        # Handle preflight OPTIONS requests
        if http_method == 'OPTIONS':
            return handle_preflight_request(event, origin=origin)
        
        if path == '/auth/register' and http_method == 'POST':
            return handle_register(event)
        elif path == '/auth/login' and http_method == 'POST':
            return handle_login(event)
        elif path == '/auth/profile' and http_method == 'GET':
            return handle_get_profile(event)
        else:
            return cors_error_response(404, 'Endpoint not found', origin)
            
    except Exception as e:
        print(f"Error in auth handler: {str(e)}")
        return cors_error_response(500, 'Internal server error', extract_origin_from_event(event))

def handle_register(event):
    """Handle user registration"""
    try:
        origin = extract_origin_from_event(event)
        body = json.loads(event['body'])
        email = body.get('email')
        password = body.get('password')
        name = body.get('name')
        role = body.get('role', 'USER')  # Default to USER role
        
        if not email or not password or not name:
            return cors_error_response(400, 'Email, password, and name are required', origin)
        
        # Create user in Cognito
        user_id = str(uuid.uuid4())
        
        cognito_response = cognito.admin_create_user(
            UserPoolId=USER_POOL_ID,
            Username=email,
            UserAttributes=[
                {
                    'Name': 'email',
                    'Value': email
                },
                {
                    'Name': 'name',
                    'Value': name
                },
                {
                    'Name': 'custom:role',
                    'Value': role
                },
                {
                    'Name': 'email_verified',
                    'Value': 'true'
                }
            ],
            TemporaryPassword=password,
            MessageAction='SUPPRESS'
        )
        
        # Set permanent password
        cognito.admin_set_user_password(
            UserPoolId=USER_POOL_ID,
            Username=email,
            Password=password,
            Permanent=True
        )
        
        # Store user data in DynamoDB
        users_table.put_item(
            Item={
                'user_id': user_id,
                'email': email,
                'name': name,
                'role': role,
                'created_at': datetime.utcnow().isoformat(),
                'updated_at': datetime.utcnow().isoformat()
            }
        )
        
        return cors_response(
            status_code=201,
            body={
                'message': 'User registered successfully',
                'user_id': user_id,
                'email': email,
                'name': name,
                'role': role
            },
            origin=origin
        )
        
    except ClientError as e:
        error_code = e.response['Error']['Code']
        if error_code == 'UsernameExistsException':
            return cors_error_response(400, 'User already exists', origin)
        else:
            print(f"Cognito error: {str(e)}")
            return cors_error_response(400, str(e), origin)
    except Exception as e:
        print(f"Registration error: {str(e)}")
        return cors_error_response(500, 'Registration failed', origin)

def handle_login(event):
    """Handle user login"""
    try:
        origin = extract_origin_from_event(event)
        body = json.loads(event['body'])
        email = body.get('email')
        password = body.get('password')
        
        if not email or not password:
            return cors_error_response(400, 'Email and password are required', origin)
        
        # Authenticate with Cognito
        response = cognito.admin_initiate_auth(
            UserPoolId=USER_POOL_ID,
            ClientId=CLIENT_ID,
            AuthFlow='ADMIN_NO_SRP_AUTH',
            AuthParameters={
                'USERNAME': email,
                'PASSWORD': password
            }
        )
        
        # Get user data from DynamoDB
        user_response = users_table.query(
            IndexName='EmailIndex',
            KeyConditionExpression='email = :email',
            ExpressionAttributeValues={':email': email}
        )
        
        user_data = None
        if user_response['Items']:
            user_data = user_response['Items'][0]
        
        return cors_response(
            status_code=200,
            body={
                'message': 'Login successful',
                'access_token': response['AuthenticationResult']['AccessToken'],
                'id_token': response['AuthenticationResult']['IdToken'],
                'refresh_token': response['AuthenticationResult']['RefreshToken'],
                'expires_in': response['AuthenticationResult']['ExpiresIn'],
                'user_data': user_data
            },
            origin=origin
        )
        
    except ClientError as e:
        error_code = e.response['Error']['Code']
        if error_code == 'NotAuthorizedException':
            return cors_error_response(401, 'Invalid credentials', origin)
        else:
            print(f"Cognito error: {str(e)}")
            return cors_error_response(400, str(e), origin)
    except Exception as e:
        print(f"Login error: {str(e)}")
        return cors_error_response(500, 'Login failed', origin)

def handle_get_profile(event):
    """Handle get user profile"""
    try:
        origin = extract_origin_from_event(event)
        # Extract user info from JWT token (passed through API Gateway authorizer)
        claims = event.get('requestContext', {}).get('authorizer', {}).get('claims', {})
        email = claims.get('email')
        
        if not email:
            return cors_error_response(401, 'Unauthorized', origin)
        
        # Get user data from DynamoDB
        user_response = users_table.query(
            IndexName='EmailIndex',
            KeyConditionExpression='email = :email',
            ExpressionAttributeValues={':email': email}
        )
        
        if not user_response['Items']:
            return cors_error_response(404, 'User not found', origin)
        
        user_data = user_response['Items'][0]
        
        return cors_response(
            status_code=200,
            body={
                'user_id': user_data['user_id'],
                'email': user_data['email'],
                'name': user_data['name'],
                'role': user_data['role'],
                'created_at': user_data['created_at']
            },
            origin=origin
        )
        
    except Exception as e:
        print(f"Profile error: {str(e)}")
        return cors_error_response(500, 'Failed to get profile', origin)

 

File: aws/src/handlers/auth_handler.js
const {
  CognitoIdentityProviderClient,
  AdminCreateUserCommand,
  AdminSetUserPasswordCommand,
  AdminInitiateAuthCommand,
  AdminListGroupsForUserCommand,
} = require("@aws-sdk/client-cognito-identity-provider");
const { DynamoDBClient } = require("@aws-sdk/client-dynamodb");
const { DynamoDBDocumentClient, PutCommand } = require("@aws-sdk/lib-dynamodb");

const cognitoClient = new CognitoIdentityProviderClient({
  region: process.env.AWS_REGION,
});

// Initialize DynamoDB client
const ddbClient = new DynamoDBClient({});
const docClient = DynamoDBDocumentClient.from(ddbClient);

// Environment variables
const USERS_TABLE = process.env.USERS_TABLE || "DogotelUsers";

exports.handler = async (event) => {
  // CORS headers - same pattern as example
  const corsHeaders = {
    "Access-Control-Allow-Origin": "*",
    "Access-Control-Allow-Headers":
      "Content-Type,X-Amz-Date,Authorization,X-Api-Key,X-Amz-Security-Token",
    "Access-Control-Allow-Methods": "POST,OPTIONS",
    "Content-Type": "application/json",
  };

  console.log("=== AUTH LAMBDA START ===");
  console.log("Event:", JSON.stringify(event, null, 2));

  // Handle preflight OPTIONS request
  if (event.httpMethod === "OPTIONS") {
    return {
      statusCode: 200,
      headers: corsHeaders,
      body: JSON.stringify({ message: "CORS preflight success" }),
    };
  }

  try {
    const path = event.path;
    const method = event.httpMethod;

    if (path === "/auth/register" && method === "POST") {
      return await handleRegister(event, corsHeaders);
    } else if (path === "/auth/login" && method === "POST") {
      return await handleLogin(event, corsHeaders);
    } else if (path === "/auth/check-admin" && method === "POST") {
      return await handleCheckAdmin(event, corsHeaders);
    } else {
      return {
        statusCode: 404,
        headers: corsHeaders,
        body: JSON.stringify({
          success: false,
          error: "Endpoint not found",
        }),
      };
    }
  } catch (error) {
    console.error("=== AUTH ERROR ===");
    console.error("Error:", error);

    return {
      statusCode: 500,
      headers: corsHeaders,
      body: JSON.stringify({
        success: false,
        error: error.message,
        message: "Authentication failed",
      }),
    };
  }
};

async function handleRegister(event, corsHeaders) {
  try {
    const body =
      typeof event.body === "string" ? JSON.parse(event.body) : event.body;
    const { email, password, firstName, lastName, birthday } = body;

    if (!email || !password || !firstName || !lastName || !birthday) {
      return {
        statusCode: 400,
        headers: corsHeaders,
        body: JSON.stringify({
          success: false,
          error:
            "Missing required fields: email, password, firstName, lastName, birthday",
        }),
      };
    }

    console.log(`Registering user: ${email}`);

    // Create user in Cognito
    const createUserParams = {
      UserPoolId: process.env.COGNITO_USER_POOL_ID,
      Username: email,
      TemporaryPassword: password,
      UserAttributes: [
        { Name: "email", Value: email },
        { Name: "given_name", Value: firstName },
        { Name: "family_name", Value: lastName },
        { Name: "birthdate", Value: birthday },
        { Name: "email_verified", Value: "true" },
      ],
      MessageAction: "SUPPRESS",
    };

    await cognitoClient.send(new AdminCreateUserCommand(createUserParams));

    // Set permanent password
    const setPasswordParams = {
      UserPoolId: process.env.COGNITO_USER_POOL_ID,
      Username: email,
      Password: password,
      Permanent: true,
    };

    await cognitoClient.send(
      new AdminSetUserPasswordCommand(setPasswordParams)
    );

    // Save user data to DynamoDB
    const userData = {
      email: email,
      firstName: firstName,
      lastName: lastName,
      username: `${firstName} ${lastName}`,
      birthdate: birthday || "", // Use birthday from signup form if provided
      profilePhoto: "", // Empty initially, can be updated in profile
      createdAt: new Date().toISOString(),
      updatedAt: new Date().toISOString(),
    };

    const putParams = {
      TableName: USERS_TABLE,
      Item: userData,
    };

    await docClient.send(new PutCommand(putParams));

    console.log("User registered successfully and saved to DynamoDB");

    return {
      statusCode: 201,
      headers: corsHeaders,
      body: JSON.stringify({
        success: true,
        message: "User registered successfully",
        email: email,
        userData: userData,
      }),
    };
  } catch (error) {
    console.error("Registration error:", error);

    if (error.name === "UsernameExistsException") {
      return {
        statusCode: 409,
        headers: corsHeaders,
        body: JSON.stringify({
          success: false,
          error: "User already exists",
        }),
      };
    }

    return {
      statusCode: 500,
      headers: corsHeaders,
      body: JSON.stringify({
        success: false,
        error: "Registration failed",
        message: error.message,
      }),
    };
  }
}

async function handleLogin(event, corsHeaders) {
  try {
    const body =
      typeof event.body === "string" ? JSON.parse(event.body) : event.body;
    const { email, password } = body;

    if (!email || !password) {
      return {
        statusCode: 400,
        headers: corsHeaders,
        body: JSON.stringify({
          success: false,
          error: "Email and password required",
        }),
      };
    }

    console.log(`Login attempt for user: ${email}`);

    const authParams = {
      UserPoolId: process.env.COGNITO_USER_POOL_ID,
      ClientId: process.env.COGNITO_CLIENT_ID,
      AuthFlow: "ADMIN_NO_SRP_AUTH",
      AuthParameters: {
        USERNAME: email,
        PASSWORD: password,
      },
    };

    const result = await cognitoClient.send(
      new AdminInitiateAuthCommand(authParams)
    );

    console.log("Login successful");

    return {
      statusCode: 200,
      headers: corsHeaders,
      body: JSON.stringify({
        success: true,
        message: "Login successful",
        accessToken: result.AuthenticationResult.AccessToken,
        idToken: result.AuthenticationResult.IdToken,
        refreshToken: result.AuthenticationResult.RefreshToken,
      }),
    };
  } catch (error) {
    console.error("Login error:", error);

    if (error.name === "NotAuthorizedException") {
      return {
        statusCode: 401,
        headers: corsHeaders,
        body: JSON.stringify({
          success: false,
          error: "Invalid credentials",
        }),
      };
    }

    return {
      statusCode: 500,
      headers: corsHeaders,
      body: JSON.stringify({
        success: false,
        error: "Login failed",
        message: error.message,
      }),
    };
  }
}

async function handleCheckAdmin(event, corsHeaders) {
  try {
    const body =
      typeof event.body === "string" ? JSON.parse(event.body) : event.body;
    const { email } = body;

    if (!email) {
      return {
        statusCode: 400,
        headers: corsHeaders,
        body: JSON.stringify({
          success: false,
          error: "Email is required",
        }),
      };
    }

    console.log(`Checking admin status for user: ${email}`);

    try {
      // Check if user is in admin group
      const listGroupsParams = {
        UserPoolId: process.env.COGNITO_USER_POOL_ID,
        Username: email,
      };

      const result = await cognitoClient.send(
        new AdminListGroupsForUserCommand(listGroupsParams)
      );
      const groups = result.Groups || [];

      // Check if user is in admin group
      const isAdmin = groups.some(
        (group) =>
          group.GroupName === "admin" ||
          group.GroupName === "Admin" ||
          group.GroupName === "ADMIN"
      );

      console.log(
        `User ${email} admin status: ${isAdmin}, groups: ${groups
          .map((g) => g.GroupName)
          .join(", ")}`
      );

      return {
        statusCode: 200,
        headers: corsHeaders,
        body: JSON.stringify({
          success: true,
          isAdmin: isAdmin,
          groups: groups.map((g) => g.GroupName),
        }),
      };
    } catch (cognitoError) {
      console.log(
        `User ${email} not found in Cognito or error checking groups:`,
        cognitoError.message
      );

      // If user doesn't exist in Cognito, check if it's a hardcoded admin email
      const adminEmails = ["admin@dogotel.com", "admin@example.com"];
      const isHardcodedAdmin = adminEmails.includes(email.toLowerCase());

      return {
        statusCode: 200,
        headers: corsHeaders,
        body: JSON.stringify({
          success: true,
          isAdmin: isHardcodedAdmin,
          groups: isHardcodedAdmin ? ["admin"] : [],
          note: isHardcodedAdmin
            ? "Hardcoded admin user"
            : "User not found in Cognito",
        }),
      };
    }
  } catch (error) {
    console.error("Admin check error:", error);

    return {
      statusCode: 500,
      headers: corsHeaders,
      body: JSON.stringify({
        success: false,
        error: "Failed to check admin status",
        message: error.message,
      }),
    };
  }
}
